pkg/roll/run_update.go

```diff
@@ -181,15 +188,56 @@ func (u *update) Run(quit <-chan struct{}) (ret bool) {
 
 	// rollout complete, clean up old RC if told to do so
 	if !u.LeaveOld {
-		u.logger.NoFields().Infoln("Cleaning up old RC")
-		if !RetryOrQuit(func() error { return u.rcStore.SetDesiredReplicas(u.OldRC, 0) }, quit, u.logger, "Could not zero old replica count") {
-			return
+		u.cleanupOldRC(quit)
+	}
+	return true // finally if we make it here, we can return true
+}
+
+func (u *update) cleanupOldRC(quit <-chan struct{}) {
+	cleanupFunc := func() error {
+		oldRC, err := u.rcStore.Get(u.OldRC)
+		if err != nil {
+			return err
 		}
-		if !RetryOrQuit(func() error { return u.rcStore.Delete(u.OldRC, false) }, quit, u.logger, "Could not delete old RC") {
-			return
+
+		if oldRC.ReplicasDesired != 0 {
+			// This likely means there was a mathematical error
+			// of some kind that was made when the RU was
+			// scheduled. If LeaveOld is false, we expect that
+			// all nodes will have been rolled from the old RC to
+			// the new one at this point, so we can delete the
+			// old RC after it isn't managing any nodes anymore.
+			// This error likely needs manual fixing by shifting
+			// the remaining nodes off of this RC and then
+			// deleting it
+			u.logger.Errorln("could not delete old RC because its replica count is nonzero")
+
+			// return nil to avoid looping, the alert
```

Comment:
i actually violated my own personal style here. usually i prefer checking the error explicitly for clarity. will fix